cmd/apigen/main.go

```diff
@@ -0,0 +1,395 @@
+/*
+Copyright 2022 The KCP Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"fmt"
+	"io/fs"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"sort"
+	"strings"
+	"time"
+
+	"github.com/bombsimon/logrusr/v3"
+	"github.com/go-logr/logr"
+	"github.com/google/go-cmp/cmp"
+	"github.com/sirupsen/logrus"
+	"github.com/spf13/pflag"
+
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	extensionsapiserver "k8s.io/apiextensions-apiserver/pkg/apiserver"
+	"k8s.io/apimachinery/pkg/api/equality"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/component-base/version"
+	"sigs.k8s.io/yaml"
+
+	"github.com/kcp-dev/kcp/pkg/apis/apis"
+	apisv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/apis/v1alpha1"
+	"github.com/kcp-dev/kcp/pkg/apis/tenancy"
+	kcpscheme "github.com/kcp-dev/kcp/pkg/client/clientset/versioned/scheme"
+)
+
+const (
+	apiResourceSchemaNamePrefix = "apiresourceschema-"
+	apiExportNamePrefix         = "apiexport-"
+)
+
+type options struct {
+	inputDir  string
+	outputDir string
+}
+
+func bindOptions(fs *pflag.FlagSet) *options {
+	o := options{}
+	fs.StringVar(&o.inputDir, "input-dir", "", "Directory containing CustomResourceDefinition YAML files.")
+	fs.StringVar(&o.outputDir, "output-dir", "", "Directory where APIResourceSchemas and APIExports will be written.")
+	return &o
+}
+
+func (o *options) Validate() error {
+	if o.inputDir == "" {
+		return fmt.Errorf("--input-dir is required")
+	}
+	if o.outputDir == "" {
+		return fmt.Errorf("--output-dir is required")
+	}
+	return nil
+}
+
+func getLogger() logr.Logger {
```

Comment:
I took this as an opportunity to try out the use of the common `logr.Logger` interface.